Game/data/building_systems/heat_system/heat_system.gd

```diff
@@ -0,0 +1,170 @@
+class_name HeatSystem
+extends Node
+## Maintains and updates the heat of all buildings.
+
+## Represents the heat flow relationships between the heat buildings.
+## Is the starting point for the Ford-Fulkerson algorithm.
+var heat_flow_graphs: Dictionary[int, HeatFlowGraph]
+
+## Represents the steady state flow of heat as the application of the Ford-Fulkerson algorithm
+## to the heat flow graph.
+## The ending state of this algorithm is as follows:
+## weight from omni-source -> source: excess heat produced by building
+## weight from source -> omni-source: consumed heat flow from sinks
+## weight from omni-sink -> sink: unused cooling from building
+## weight from sink -> omni-sink: consumed heat flow from sources
+var steady_state_flows: Dictionary[int, HeatFlowGraph]
+
+## List of all heat sources.
+var heat_sources: Dictionary[int, Array]
+
+## List of all heat sinks.
+var heat_sinks: Dictionary[int, Array]
+
+func _ready() -> void:
+	ComponentManager.component_added.connect(_on_component_added)
+	ComponentManager.component_removed.connect(_on_component_removed)
+	Model.game_ready.connect(_on_game_ready)
+
+
+## When a heat component is added, update the internal state and re-calculate the steady state
+## flow of heat.
+func _on_component_added(component: BuildingComponent) -> void:
+	if component is HeatComponent:
+		var player_id: int = component.building_entity.player_id
+
+		heat_flow_graphs[player_id].add_building(component)
+
+		if component.is_source:
+			heat_sources[player_id].append(component)
+		elif component.is_sink:
+			heat_sinks[player_id].append(component)
+
+		Model.add_heat_data_at.rpc(
+			component.building_entity.player_id,
+			component.building_entity.position,
+			component.heat,
+			component.heat_capacity)
+
+		calculate_steady_state_flow(player_id)
+
+		print_flow_rates()
+
+
+## When a heat component is removed, update the internal state and re-calculate the steady state
+## flow of heat.
+func _on_component_removed(component: BuildingComponent) -> void:
+	if component is HeatComponent:
```

Comment:
Again consider the "early out" approach to this instead of putting the entire function under a conditional